### ISSUE TITLE
Offset infogrid / jitter using X, without color

### ISSUE THREAD

**Issue by sharlagelfand** [microsoft/datamations#61]
Seems like infogrid / jitter case can only handle when the offsetting variable is color:

```r
"small_salary %>%
    group_by(Degree, Work) %>%
    summarize(mean_salary = mean(Salary, na.rm = TRUE)) %>%
    ggplot(aes(x = Degree, y = mean_salary)) +
    geom_point() +
    facet_grid(~ Work)" %>%
  datamation_sanddance() 
```

Specs [here](https://github.com/microsoft/datamations/blob/parse-ggplot2/sandbox/ggplot2/offset_x_with_color.json).

https://user-images.githubusercontent.com/15895337/119526361-528c7400-bd4d-11eb-85df-d26b0acad222.mov

The above R code says nothing about color, and really should only be offsetting by X, but this doesn't work (I manually added color in to the specs above)

https://user-images.githubusercontent.com/15895337/119526824-badb5580-bd4d-11eb-82f3-93086eeb5d1e.mov

Specs [here](https://github.com/microsoft/datamations/blob/parse-ggplot2/sandbox/ggplot2/offset_x_only.json). And here is the [jitter](https://vega.github.io/editor/#/url/vega-lite/N4IgJAzgxgFgpgWwIYgFwhgF0wBwqgegIDc4BzJAOjIEtMYBXAI0poHsDp5kTykBaADZ04JAKyUAVhDYA7EABoQAEySYUqUMSSCGcCGgDaoMohqyaAfRrK0ARiUB1NgCcA1mhABBKEmVmUJQARchc4OE8AWSQITDgXAyUAD3slAE80AA47SgBOABZ8sQAGOwBmTLKANlyxTIBfBRMzC2tbVAAmJ1cPdB8-AMUQELIwiPRo2PjEkBTUBxAM1EyOykzasuLisSr8lbsGppBTBHMrGzQy7vdPfv9TwOHQ8KiYuIShuYWlysoOqp2dgA7B18iDth1Gs1Tq0Lqh8tdet5fPcaI8RmNXlMPslUossqt1rUwf8Afk7FDji1zu0xIjbijBsFnuMQJN3jMvuksvlKGUOoCQUC6rlcmVKSczm00FV6X1GQ8hhiXhM3tNPnifry7PlilVKnYBR0KrkJdTpaggXLkQNFczRiq2WqcbNNVkJGVNvzSmV8qKDmaYTSsta7kyng7Wez1bj5tzlhI-WIdQdNqKKoGpXDcqGFWilSysRyNXH8cteWJdZkxNU-VU7MUgZnYe0G7nbfn7ZjVdjOW6E5R8vXQQV1plih1TUdJS37AtnDd5R30YX0AAFGBBEtdMuZXl7fWespiY1VDpN6fmuGG9uoleRzwbrexnc-MprIe5cq5TJ7mqQy8gwtcpb3DZVWSfbd41+YUynKfYxCBIcL2hLNWwREAFyRMM7QjbsQEgl9oIkf5KlBfI4P+PZm2DeY6UwnoGWXAsH3XTcoN3D002BXZciqGoqho4DZQYxcbTvFj8MI11Omg3kgQqIcwVFMF9SE68rVE7C83vKT2KIzjKC2fUxBPDpik2OwKUAtD7EyUDcPAx99Jk193UoOwqk2PV9R-AFrNQ2d5hzLSmIkrtHWkuY3IHMogQUoFPJ8v1DkC2jzIczs8MilzouIoyVl-IoG2TX11PaDp50YpdwuyiDcrQGLMiqIya2rMpcgFYo9xQqkgLhDodywsKwNXAiGtkwyvz1IpRRHZNysaq5Qpq0bWPG59XOglril1LzjzsStK1SvrbM6DDhtWxyxqixrtsHQoQRWYov0S-JFs6ejLvEta9M2vLdxa6ogV-TIgQnRtcl6md0pE76cKypy2P+u7dyBIy-QnL8Oma-4PvPTLdJylHJp+Fr9XPcpK02Xj8fslafuu9bbtJrJ0Yo16CiQgUFpsoLJ2tABJWRlAYWIXAyCKo2dPtSyWXJig8sGgS-coqhVuwpzSi1NiFkWxcwCXJMdaMXS5MsFb5eCFaPWCAtOoK4L10XxcluqixjGTvjQS3PQBYVEL9XV7ZhnWhuqkBhZdw23aRp1exLb3llySh9WrQ6QeNHZxT52jPWdg2jalj2zf7dZU8QjrNYohTBNznWLojqPC9jsbTdlpPy+yCyDke8l+I+48C9d43pYT2Mk8tnUihKUrSm2Qe4ab-WR+Lntiwn+NLYFKGtknc9Kve+u4Ti4eY9HkuO+glO6kbc9dl1OCtYdvP6e+5vV-d9fPfN+WcjBYUNRKxeWBNDK87QOpnyLl-eOG8vZb0VpWQ+Wxthnkqh9XUUDW7rXbonBBeQSgrBrKUMUFkMFVTEh-c+a9YE-37JbIEVRuqDT3MmRCAFtZwnyOHShK9qEwNwZvC2ORD61F-DsW+J1Q5cOWu-Ph0C46CPgcIwc2RDpiFFO1TqGDG68OjgotuMs8EqPMsebYYMrKawwV9Ze+jsH4SUb-H2isFZbBUurRKz9pHtCHFgi+39S5yx9jkX0YITylCsoNDhL8LRgj8TQxx9DFb1n-EOY0fo4IYLfrYlu-jaGBMnjkYEP4ayUQonuDBIU5F2LyYkoJqAvx-AyUhCmcEB7H1pMUeJAijFCL-oOXaYo4rg3EWA-qtIKFIioQYnBvTlH9KsgCY8vpfygI+iebpii5lOIaUUvUzVuHpk6jnThtJZE5M-ls8e8znGp3Btw8cwNDQh3AWgSsmzDHXJ2Y0z8f51bTxBOsmxejckJO2fQnI2wVaoM8vco+py3lLxBZcz5cDvkuIooUTyk4wamXWZpapoKelfIhR5JC2xEITmFIddZ2TkX8KuWi0loJShmWqDWUU6yqkXIZaiuh9TLYlCshUMi4ipGvNQEwj5sySUCuSbtLGvoBT8mid4mUkzPDTPsSbcFAr-5bBKJ+UEiE64IslTwqZ8jtVjyZXqjyqlRR8RegpF54yZTnPpTMhxurJ6K3uXUEGCEAwdJlLoy1NSwWyt9anA4lRPE9UrB9AE0rvVRvwdnA6FlNiDVNTEuE-EU06rTRbRBP5NbU0Qs8pNBKeVeqLbawpZKti+gnOrSmSa6XhqJYy-l0akJPRVjWBS2Qk3cs9day+xj5arFKPWHUINTJpw+uDQtNre1b1WDjauyyVZinhXm9oiVV2Tr6cEyg4TDoHGNHqPUy6LWaqtbUn1G6ml6mNKSBWwazUKWPQEq+KiUk6jKbsOKy6w0PojcShtL7do6hqHFQoFQxlnWFL+-J-7p1-E6jqH8+ymHirdZaJFXaUUyugwBstyZBo-nHARlDNbx1PuLf09YTDVJ+nMu079naIPdr5QUl95JlaVX4hov0y6x0kd5WR9dKiwb8RdSDXav4PrjjQ3UyeqwKi7X5JWRscbVMavQFqpj5HMPAknOrDRg1qaqfvcZx9kazM+1WIdFhLCiQAlUx6qTda10CYtq5goi6XqNmUqp8DDnIM9oC-00yUMDh8U42QkNCZ1PPpUXNSqO8caIVU8R3jpHU3OYaasIc1ZyT7zBKB1LYN0vMbPT+Tqc0djPVVRK389WSuNNFNnVS-Il21ck4V6TxXZOYc2HuasNRPw1g+grLr42z1IWPIlMTKx+TzaM5HRzMCWYxR+RCKyaiDjq3m-Znb0WboTQO1pvYBRuoWQolxg9PsfMjb885EmB33xJX+CJ8kBRXVnQKGh-bW93y6hKaZCc6xdrzeBb5idyMOLy3fFCt94MRWenmwVqLfHmY3ZfQpCo5lmk-mQ0FKGYOicW1+0wl6nkvwAgXqln8NPvsQ8oFj7Y6ZqxPfm8N-HRXiao7e2sT08VBrYoFB9BsXSGYmZoeDi2vIJyXsNDsGbDQAC6jQQAADMkBQDgJgNAoAoBsF0AgeQmhDc0DgIIdooUlCYDSDgVkrhlDmB0EMTAdBBCsmGvUfXEAPdQHNxgOANAyBYEuFsJQAB3Gw9B7AlCUMgRcoA3ce88DgNg5gzdKANzQQQgf2iGz0PruAshLfe9kGQSPcxQAl8d87lIrv3esgAI4MCQLIf36h-ekCGEgJINADB2+0LofQRgHAdB10oQQSAmCO4AKJJBwC4TwLg2AMBFgAClUJgBgCBKDL9X4IAAlAAAgALx35v3YG-AB+G-ABydu7+b+oA-0+d-IA+ug+gengTkSg0AOg4woAygbAyA5gRgiCCgqwYgeu+uSwLeDuTungbsOePefeA+dAagNAI+ruAerIAAyjoEgNAhASAXbjAXAbboYNkHkIUEKhUNUGIgoAUEZJVMmJrvxKOKgSHvUEAA) frame to show how it should look (the summary frame is correct, but jitter is not, neither is infogrid).

cc @giorgi-ghviniashvili

**Comment by giorgi-ghviniashvili** [comment]
@sharlagelfand yes, that's right, it only shifts on color. Could you please include jitterField in the meta? For example, for color, meta will be: { parse: "jitter", jitterField: "sex" }, for degree split (current frame), it will be { parse: "jitter", jitterField: "x" }. If jitterField is not available or is null, it will not split points and will treat like a single group. 

Here is example if jitterField is x:

![image](https://user-images.githubusercontent.com/6615532/119827561-56192a00-bf0a-11eb-819d-1c211d690134.png)

**Comment by giorgi-ghviniashvili** [comment]
Or maybe let's call it `splitField` or `offsetField`?

**Comment by giorgi-ghviniashvili** [comment]
Ok, using `splitField` for both `parse:jitter` and `parse:grid`.

**Comment by sharlagelfand** [comment]
Thanks @giorgi-ghviniashvili! This works well for this first example:

https://user-images.githubusercontent.com/15895337/119843642-bcd02080-bed5-11eb-801a-e2de4a097681.mov

Though I realized that now, without the colour, in the third frame (~5 seconds in) we have nothing to distinguish masters from phd, might be something for us to keep in mind re: adding an axis in to distinguish (cc @jhofman).

I'm looking back on the other example from #40 [here](https://github.com/microsoft/datamations/issues/40#issuecomment-844404361), and I think the second row of the grid needs to be offset so that it matches up with the jitter in the next frame:

```r
"small_salary %>%
  group_by(Degree, Work) %>%
  summarize(mean_salary = mean(Salary, na.rm = TRUE)) %>%
  ggplot(aes(x = Degree, y = mean_salary)) +
  geom_point() +
  facet_grid(Degree ~ Work)" %>%
  datamation_sanddance()
```

https://user-images.githubusercontent.com/15895337/119844516-7e873100-bed6-11eb-857a-74951f5357c9.mov

[Here](https://github.com/microsoft/datamations/blob/parse-ggplot2/sandbox/ggplot2/offset_x_second_row.json) are the specs for that.

**Comment by giorgi-ghviniashvili** [comment]
@sharlagelfand  about the third frame, yes we will need x-axis and then instead of this: 

![image](https://user-images.githubusercontent.com/6615532/119877720-6e08a200-bf3a-11eb-9bfa-3b5455b7c875.png)

We can use this (which is already offset and matches next jitter frame): 

![image](https://user-images.githubusercontent.com/6615532/119877744-74971980-bf3a-11eb-8ffa-c41c972a14fa.png)

**Comment by sharlagelfand** [comment]
@giorgi-ghviniashvili the goal is to still have both column and row facets, offset by X, not to translate into column facets and x values only. We want a user to be able to specify the final form so if they specify col + row facets, with X offset, we should render that

**Comment by giorgi-ghviniashvili** [comment]
@sharlagelfand so do you mean in this frame grid need to be offseted? Can we first add x-axis?

![image](https://user-images.githubusercontent.com/6615532/119993069-3c481780-bfdc-11eb-92b8-03b5de410f9a.png)

**Comment by sharlagelfand** [comment]
Yes, exactly @giorgi-ghviniashvili 

The grid looks like this

<img width="553" alt="Screen Shot 2021-05-28 at 10 41 47 AM" src="https://user-images.githubusercontent.com/15895337/120001024-6203fb00-bfa1-11eb-9650-df304591d01d.png">


and the next frame looks like this

<img width="593" alt="Screen Shot 2021-05-28 at 10 42 00 AM" src="https://user-images.githubusercontent.com/15895337/120001052-66c8af00-bfa1-11eb-9f83-0762a870dcf6.png">


which is confusing, because the transition looks like this

<img width="606" alt="Screen Shot 2021-05-28 at 10 42 10 AM" src="https://user-images.githubusercontent.com/15895337/120001066-68927280-bfa1-11eb-814a-8437182550c8.png">

so the first frame needs to be offset as well so that the line up in the transition and are in the correct place.

re: adding axes, that was my question in [this comment](https://github.com/microsoft/datamations/issues/61#issuecomment-849686012) about whether it might make sense to have (just x) axes in the grid as well.

**Comment by giorgi-ghviniashvili** [comment]
@sharlagelfand  that's already achievable.

In third frame [here](https://raw.githubusercontent.com/microsoft/datamations/parse-ggplot2/sandbox/ggplot2/offset_x_second_row.json), please remove `splitField` (because you are not actually splitting within groups). SplitField should only be used when you want to further split a single grid specified in `data.values` with `n`.

And to shift it, please modify `data.values` like this:

![image](https://user-images.githubusercontent.com/6615532/120077300-e0a08b80-c0ba-11eb-8a9d-012d9ffaef28.png)

(70 means shift by 70 grid points, which will be 7 columns).
(90  = 90 points, 9 columns).

"So a single column consists of 10 points.."
----
It will render this:
![image](https://user-images.githubusercontent.com/6615532/120077400-573d8900-c0bb-11eb-805c-62a8f3770f64.png)
----
Let me know if that works.

**Comment by giorgi-ghviniashvili** [comment]
https://user-images.githubusercontent.com/6615532/120077510-c74c0f00-c0bb-11eb-9b5b-27e1e9d736e3.mov

**Comment by sharlagelfand** [comment]
@giorgi-ghviniashvili I'm hesitant to introduce any `shiftX` logic into the R side, since it's all being calculated on the JS side right now - we would be duplicating all the logic that already exists for calculating this into R and if there are any changes (e.g. if we want to change how many points are in a column) it would have to be done in two places. What do you think about handling adding that in on your end?

There is another case that I need to think about for this, e.g. if Work is in column facets, Degree is in row facets, and Work is on the X axis, how it should look. I'll put together some specs / a sketch for this

**Comment by sharlagelfand** [comment]
Here is a [sample spec](https://github.com/microsoft/datamations/blob/parse-ggplot2/sandbox/ggplot2/offset_x_second_column.json) for the example I described above, where Work is in column facets, Degree is in row facets, and Work is on the x-axis, created with this code:

```r
small_salary %>%
  group_by(Degree, Work) %>%
  summarize(mean_salary = mean(Salary, na.rm = TRUE)) %>%
  ggplot(aes(x = Work, y = mean_salary)) +
  geom_point() +
  facet_grid(Degree ~ Work)
```

Here are what the frames look like:

<img width="796" alt="Screen Shot 2021-05-31 at 11 06 43 AM" src="https://user-images.githubusercontent.com/15895337/120212886-9295a000-c200-11eb-9123-5305775f8ccd.png">

<img width="830" alt="Screen Shot 2021-05-31 at 11 06 48 AM" src="https://user-images.githubusercontent.com/15895337/120212891-945f6380-c200-11eb-8cfb-52eacf369e8c.png">'

With the animation between them looking like this:

<img width="834" alt="Screen Shot 2021-05-31 at 11 06 59 AM" src="https://user-images.githubusercontent.com/15895337/120212916-9b867180-c200-11eb-972f-52b547211d1d.png">

Ideally, to be consistent with the final plot, the infogrid for the right column should have those points "offset" by the width of the points in the first column

**Comment by sharlagelfand** [comment]
Not sure if this is related to this (I'm losing track a bit of which bugs fit where!) but [this spec]() has its X offset for grid working great:

<img width="609" alt="Screen Shot 2021-05-31 at 1 54 36 PM" src="https://user-images.githubusercontent.com/15895337/120227739-c03a1380-c217-11eb-951a-b9d99066e3ae.png">


but the axes on the jitter is off

<img width="675" alt="Screen Shot 2021-05-31 at 1 50 46 PM" src="https://user-images.githubusercontent.com/15895337/120227745-c29c6d80-c217-11eb-84b1-897d5f822f1e.png">

The spec in its own looks fine, and in the next frames, just something happening with the axes I think (meta.axes is set to FALSE)

<img width="403" alt="Screen Shot 2021-05-31 at 1 51 17 PM" src="https://user-images.githubusercontent.com/15895337/120227747-c4663100-c217-11eb-8c68-7c8d1858b43a.png">

**Comment by giorgi-ghviniashvili** [comment]
@sharlagelfand  please update link of "this spec" , it does not link anything.

**Comment by giorgi-ghviniashvili** [comment]
@sharlagelfand  agree that it should be done in JS side, let me think a bit how we can solve this, most likely we will need a new field in meta to find out when should js do it.

**Comment by sharlagelfand** [comment]
Oops sorry @giorgi-ghviniashvili! Spec updated in that comment now

**Comment by giorgi-ghviniashvili** [comment]
Ok, that works because we don't have facets and grid, we only have one group of data which is **split by** spitField and jittered. Works great.

Because jitter, `domain` is adjusted while `axis` stays same, so I will need ['Academia', 'Industry'] to be added to somewhere to be able to correctly position them. 

![image](https://user-images.githubusercontent.com/6615532/120230711-fb146980-c260-11eb-980b-c70c55e419c4.png)

For Academia, I will set it to 75 and for Industry I will set it to 225. NO HARDCODE.

P.S. I could get it using splitField, but can't make it correctly sorted.

**Comment by giorgi-ghviniashvili** [comment]
![image](https://user-images.githubusercontent.com/6615532/120231548-affb5600-c262-11eb-855e-03ee2d010089.png)

**Comment by giorgi-ghviniashvili** [comment]
We can add it as `meta.xAxisLabels`.

**Comment by giorgi-ghviniashvili** [comment]
> There is another case that I need to think about for this, e.g. if Work is in column facets, Degree is in row facets, and Work is on the X axis, how it should look. I'll put together some specs / a sketch for this

I think having column facets and x axis is same fields is very confusing, because In Academia, you have Academia and Industry, same for Industry as facet.  In general, I think Column facet field should not match x axis field.

**Comment by giorgi-ghviniashvili** [comment]
And about this one:

![image](https://user-images.githubusercontent.com/6615532/120233447-df13c680-c266-11eb-86dd-aa0574aa72dd.png)

When you add `meta.xAxisLabels`, I will try to add `shiftX` myself matching `xAxisLabels` array. 

For this to know when to do please add `meta.shiftGrids: true` (naming suggestions welcome).

**Comment by sharlagelfand** [comment]
> We can add it as `meta.xAxisLabels`, done it [here](https://github.com/microsoft/datamations/commit/dc9e66b8bc4e8e429fdc73313bb7a53872997484#diff-0ffac8256451042e821e4bb2a1248e3b5cd158e844b8fa0cc3ab20abb9f6bf78R209).

This info is already contained in the labelExpr - is the reason it can't be pulled from there because of the ordering?


> I think having column facets and x axis is same fields is very confusing, because In Academia, you have Academia and Industry, same for Industry as facet. In general, I think Column facet field should not match x axis field.

I agree, I think it looks pretty bad and confusing :) I don't necessarily want to encourage it, but if we want to take ggplot2 code as input to specify how the plot looks (#40) , then that's a case that should be handled

**Comment by giorgi-ghviniashvili** [comment]
> This info is already contained in the labelExpr - is the reason it can't be pulled from there because of the ordering?

No, I don't want parse the expression.. What if we need another expression or expression is not correct at all? Better to put it in meta and also use it for shift to match facet with x-axis.

**Comment by sharlagelfand** [comment]
Fair! It's probably easier to pull it out in the correct order in R and just pass it along.

To confirm... meta.xAxisLabels should be set in the jitter spec when parse: jitter, splitField: true, axes: false (if axes: true it doesn't need to be set, because these are hacked facets which already handles all of the x axis labelling)

and then also set in cases where the infogrid needs to be split, so parse: grid, splitField: x, and shiftGrids: true also include meta.xAxisLabels (both cases when x is the same as column facet or when x is the same as row facet)?

Once we have this all sorted... I will make a table of all the meta settings and when they are flagged / how to use them :)

**Comment by giorgi-ghviniashvili** [comment]
> meta.xAxisLabels should be set in the jitter spec when parse: jitter, splitField: true, axes: false (if axes: true it doesn't need to be set, because these are hacked facets which already handles all of the x axis labelling)

Yes, correct!

> and then also set in cases where the infogrid needs to be split, so parse: grid, splitField: x, and shiftGrids: true also include meta.xAxisLabels (both cases when x is the same as column facet or when x is the same as row facet)?

In this case, please remove splitField, because in case of info-grid, we are not actually splitting anything.

> Once we have this all sorted... I will make a table of all the meta settings and when they are flagged / how to use them :)

Yes, we need it for sure

**Comment by sharlagelfand** [comment]
Ok awesome - I will update it hopefully later today or tomorrow!

For the splitField in info grid - do we still need it in the cases when it _is_ split? If the grid is split by a variable in `color` for example

**Comment by sharlagelfand** [comment]
Thanks @giorgi-ghviniashvili, adding `xAxisLabels` works great for this example:


https://user-images.githubusercontent.com/15895337/120347286-24b5ab00-c2ca-11eb-84f3-46af377dfa60.mov



I'm looking at adding shiftGrids in, it hasn't been added in on your end right? [Here](https://github.com/microsoft/datamations/blob/parse-ggplot2/sandbox/ggplot2/shiftgrids_example.json) is a spec that contains it that we can test with.

**Comment by giorgi-ghviniashvili** [comment]
@sharlagelfand  can we please start with a situation when column facet and x-axis are different? In [this](https://raw.githubusercontent.com/microsoft/datamations/parse-ggplot2/sandbox/ggplot2/shiftgrids_example.json) spec, they both are same and it is ambiguous for me

**Comment by jhofman** [comment]
We realized we were in the weeds on edge cases here, so let's settle some conventions in #68 first before continuing on this.

@giorgi-ghviniashvili: side notes are that we'd like some more labeling of infogrid groups when there's a non-obvious split (by perhaps turning on x axis ticks), and maybe we should always center-justify infogrids to match the center-justification of jitter plots in following frames, so that center justify is the new default?

**Comment by giorgi-ghviniashvili** [comment]
@jhofman I think that this one should be closed, since we continue working in #68.

**Comment by jhofman** [comment]
sounds good!
